crates/sui-indexer/src/handlers/mod.rs

```diff
@@ -44,3 +52,104 @@ pub struct EpochToCommit {
+
+        let mut unprocessed = BTreeMap::new();
+        let mut tuple_batch = vec![];
+        let mut next_cp_to_process = self
+            .handler
+            .get_watermark_hi()
+            .await?
+            .map(|n| n.saturating_add(1))
+            .unwrap_or_default();
+
+        loop {
+            if cancel.is_cancelled() {
+                return Ok(());
+            }
+
+            // Try to fetch new data tuple from the stream
+            match stream.next().now_or_never() {
+                Some(Some(tuple_chunk)) => {
+                    if cancel.is_cancelled() {
+                        return Ok(());
+                    }
+                    for tuple in tuple_chunk {
+                        unprocessed.insert(tuple.0, tuple);
+                    }
+                }
+                Some(None) => break, // Stream has ended
+                None => {}           // No new data tuple available right now
+            }
+
+            // Process unprocessed checkpoints, even no new checkpoints from stream
+            let checkpoint_lag_limiter = self.handler.get_checkpoint_lag_limiter().await?;
+            while next_cp_to_process <= checkpoint_lag_limiter {
+                if let Some(data_tuple) = unprocessed.remove(&next_cp_to_process) {
+                    tuple_batch.push(data_tuple);
+                    next_cp_to_process += 1;
+                } else {
+                    break;
+                }
+            }
+
+            if !tuple_batch.is_empty() {
+                let last_checkpoint_seq = tuple_batch.last().unwrap().0;
+                let batch = tuple_batch.into_iter().map(|t| t.1).collect();
+                self.handler.load(batch).await.map_err(|e| {
+                    IndexerError::PostgresWriteError(format!(
+                        "Failed to load transformed data into DB for handler {}: {}",
+                        self.handler.name(),
+                        e
+                    ))
+                })?;
+                self.handler.set_watermark_hi(last_checkpoint_seq).await?;
+                tuple_batch = vec![];
+            }
+        }
+        Err(IndexerError::ChannelClosed(format!(
+            "Checkpoint channel is closed unexpectedly for handler {}",
+            self.handler.name()
+        )))
+    }
+}
+
+#[async_trait]
+pub trait Handler<T>: Send + Sync {
+    /// return handler name
+    fn name(&self) -> String;
+
+    /// commit batch of transformed data to DB
+    async fn load(&self, batch: Vec<T>) -> IndexerResult<()>;
+
+    /// read high watermark of the table DB
+    async fn get_watermark_hi(&self) -> IndexerResult<Option<u64>>;
+
+    /// set high watermark of the table DB, also update metrics.
+    async fn set_watermark_hi(&self, watermark_hi: u64) -> IndexerResult<()>;
+
+    /// get checkpoint lag limiter, for handlers without lag limiter, return u64::MAX
+    async fn get_checkpoint_lag_limiter(&self) -> IndexerResult<u64> {
```

Comment:
The function name is not very intuitive. It seems what you are returning is really the max_committable_checkpoint?
And a bit more detailed doc comment would be great too.